File: src/Engine/Core/KeyFrame.cpp
#include "KeyFrame.h"
#include <iostream>

//public:
//    //--------------
//    // -- FIELDS --
//    //--------------
//    std::vector<Point16> actionPoints;
//    Vector3D* impulse;      // if it's mooving
//    Vector3D* throwImpulse; // give impulse when throwing someone
//    UInt index;             // UByte
//    UInt actionType;        // UByte
//    UInt throwDamage;       // UByte
const int KeyFrame::INVALID_VEL = -128;
const int KeyFrame::NO_CHANGE_VEL = -127;

const UByte KeyFrame::FRONTAL_KNOCK = 1;
const UByte KeyFrame::SPARSE_KNOCK = 2;
const UByte KeyFrame::BACK_KNOCK = 4;

//--------------------
// -- CONSTRUCTORS --
//--------------------
KeyFrame::KeyFrame(DataInputStream dis) throw(IOException){
    readData(dis);
}
KeyFrame::KeyFrame():HitFrame(), impulse(0),index(0)
{}

//---------------
// -- METHODS --
//---------------

void KeyFrame::readData(DataInputStream& dis) throw(IOException){
    if (throwImpulse){ delete throwImpulse; throwImpulse = 0; }
    actionPoints.clear();

    index = dis.readByte();
    // action
    if (actionType != HitFrame::NO_ACTION){
        actionType = dis.readByte();
        if (actionType >= GameInvariants::AT_KNOCK){
             UByte knockFlags= dis.readByte();
             frontalKnock = knockFlags & FRONTAL_KNOCK;
             backKnock = knockFlags & BACK_KNOCK;
             sparseKnock = knockFlags & SPARSE_KNOCK;
        }

        // poins
        int n = dis.readByte();
        actionPoints.reserve(n);
        for(int i=0; i<n; ++i)
            actionPoints.push_back(Point16(dis));
        // throw stuff
        if (actionType == GameInvariants::AT_THROW){
            throwImpulse = new FloatVector3D();
            throwImpulse->readSignedByteData(dis);
            //std::cout << throwImpulse->x << ", " << throwImpulse->y << ", " << throwImpulse->z << "\n";
            throwImpulse->x/=GameInvariants::VEL_DIVIDER;
            throwImpulse->y/=GameInvariants::VEL_DIVIDER;
            throwImpulse->z/=GameInvariants::VEL_DIVIDER;
            throwDamage = dis.readByte();
            throwRot = dis.readSignedByte();
            throwRot/=GameInvariants::ROT_VEL_DIVIDER;
        }
    }
    // movement impulse
    if (impulse){
        impulse->x = dis.readSignedByte();
        if (impulse->x!=INVALID_VEL){
            impulse->y = dis.readSignedByte();
            impulse->z = dis.readSignedByte();

            if (impulse->x != NO_CHANGE_VEL)impulse->x/=GameInvariants::VEL_DIVIDER;
            if (impulse->y != NO_CHANGE_VEL)impulse->y/=GameInvariants::VEL_DIVIDER;
        }
        //else impulse->x = impulse->y = impulse->z = 0;
    }
}


//void KeyFrame::writeData(DataOutputStream& dos){ //throws IOException
//    index.writeData(dos);
//    // action
//    if (actionType!=null){
//        actionType.writeData(dos);
//        // points
//        dos.write(actionPoints.size());
//        for(IntPoint p:actionPoints){
//            p.writeData(dos);
//        }
//        // throw stuff
//        if (actionType.get() == game.GameInvariants.AT_THROW && throwImpulse != null){
//            throwImpulse.writeData(dos);
//            throwDamage.writeData(dos);
//        }
//    }
//    // movement impulse
//    if (impulsed){
//        int toWrite = impulse.x;
//        if (impulse.x == 0 && impulse.y == 0 && impulse.z == 0)
//            toWrite = INVALID_VEL;
//        dos.writeByte(toWrite);
//        if (toWrite!=INVALID_VEL){
//            dos.writeByte(impulse.y);
//            dos.writeByte(impulse.z);
//        }
//    }
//}

File: src/Editors/Editors Lib/src/characters/KeyFrame.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package characters;

import file.IReadWrite;
import game.*;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import util.BTriple;
import util.IntPoint;
import util.UBPoint;
import util.UByte;

/**
 *
 * @author Gil
 */
public class KeyFrame implements IReadWrite{
    public static final byte INVALID_VEL = Byte.MIN_VALUE;
    
    public static final int FRONTAL_KNOCK = 1;
    public static final int SPARSE_KNOCK = 2;
    public static final int BACK_KNOCK = 4;
    //--------------
    // -- FIELDS --
    //--------------
    public int versionID;
    
    public UByte index;
    public UByte actionType;
    public List<IntPoint> actionPoints;
    public BTriple impulse;
    public BTriple throwImpulse;
    public UByte throwDamage;
    public byte throwRotation;
    public boolean impulsed;
    public boolean invalid;
    
    // Knocking flags
    public UByte knockFlags;
    
    //--------------------
    // -- CONSTRUCTORS --
    //--------------------

    public KeyFrame(DataInputStream dis, int version) throws IOException {
        versionID = version;
        this.readData(dis);
    }
    public KeyFrame(int version){
        versionID = version;
        this.actionPoints = new ArrayList<IntPoint>();
        index = new UByte();
        actionType = new UByte();
        impulse = new BTriple();
        knockFlags = new UByte();
    }
    public KeyFrame(KeyFrame other){
        this.versionID = other.versionID;
        this.impulsed = other.impulsed;
        this.index = other.index;
        this.actionType = other.actionType;
        this.knockFlags = other.knockFlags;
        this.impulse = new BTriple(other.impulse.x, other.impulse.y, other.impulse.z);
        this.actionPoints = new ArrayList<IntPoint>(other.actionPoints.size());
        for(IntPoint p:other.actionPoints)
            actionPoints.add(p.clone());
    }
    

    //---------------
    // -- METHODS --
    //---------------
    
    protected void writeVs1_2_3(DataOutputStream dos) throws IOException {
        index.writeData(dos);
        // action
        if (actionType!=null){
            actionType.writeData(dos);
            if (actionType.get()>=GameInvariants.AT_KNOCK && versionID>=3)
                knockFlags.writeData(dos);
            // points
            dos.write(actionPoints.size());
            for(IntPoint p:actionPoints){
                p.writeData(dos);
            }
            // throw stuff
            if (actionType.get() == game.GameInvariants.AT_THROW && throwImpulse != null){
                throwImpulse.writeData(dos);
                throwDamage.writeData(dos);
                if (versionID>=2) dos.writeByte(throwRotation);
            }
        }
        // movement impulse
        if (impulsed){
            int toWrite = impulse.x;
            if (impulse.x == KeyFrame.INVALID_VEL)
                toWrite = INVALID_VEL;
            dos.writeByte(toWrite);
            if (toWrite!=INVALID_VEL){
                dos.writeByte(impulse.y);
                dos.writeByte(impulse.z);
            }
        }
    }

    protected void readVs1_2_3(DataInputStream dis) throws IOException {
        index = new UByte(dis);
        // action
        if (actionType !=null){
            actionType.readData(dis);
            if (actionType.get()>=GameInvariants.AT_KNOCK && versionID>=3)
                knockFlags = new UByte(dis);
            // poins
            int n = dis.readByte();
            actionPoints = new ArrayList<IntPoint>(n);
            for(int i=0; i<n; i++)
                actionPoints.add(new IntPoint(dis));
            // throw stuff
            if (actionType.get() == game.GameInvariants.AT_THROW){
                throwImpulse = new BTriple(dis);
                throwDamage = new UByte(dis);
                if (versionID>=2) throwRotation = dis.readByte();
            }
        }
        // movement impulse
        if (impulsed){
            impulse.x = dis.readByte();
            if (impulse.x!=INVALID_VEL){
                impulse.y = dis.readByte();
                impulse.z = dis.readByte();
            }
            else impulse.y = impulse.z = 0;
        }
    }
    
    
    
    @Override
    public void writeData(DataOutputStream dos) throws IOException {
        switch(versionID){
            case 1: writeVs1_2_3(dos); break;
            default: writeVs1_2_3(dos);
        }
    }
     
    @Override
    public void readData(DataInputStream dis) throws IOException {
        switch(versionID){
            case 1: readVs1_2_3(dis); break;
            default: readVs1_2_3(dis);
        }
    }
    
    
}
